### ISSUE TITLE
Someday we will have time to build a lovely Go multi-everything router.. :_(

### ISSUE THREAD

**Issue by nerdalert** [gopher-net/gopher-net#1]
or maybe lets take up digital gold farming.

**Comment by Mierdin** [comment]
I'm so close to having time to work on at least the design of this...we just all need to sit down and hash out some ideas.

**Comment by dave-tucker** [comment]
> we just all need to sit down and hash out some ideas.

In the same place? I'm not sure one room would be sufficient to contain that much awesome

**Comment by Mierdin** [comment]
Tonight I said f**\* it and just started looking at inter-process communication methods in go. Any pointers? At this point I'm leaning towards simple [RPCs over sockets](http://golang.org/pkg/net/rpc/).

**Comment by dave-tucker** [comment]
Do we need IPC at all? I'd have thought we would have a single daemon, it can spawn a process for each protocol it needs to run and using channels for inter-goroutine comms.

**Comment by Mierdin** [comment]
I really believe the design laid out in Quagga is the right way to go - limiting potential failure domains by running important processes as truly separate processes. A failure in BGP wouldn't necessarily bring down the OSPF process, etc. Plus, we'd need a watchdog process anyways.

[The way it's laid out in Quagga](http://www.nongnu.org/quagga/index.html) is that the watchdog process (zebra) is also the abstraction layer between the routing daemons and the kernel. So the daemons don't need to interact with the kernel directly, they make calls to zebra using a [simple protocol](http://www.nongnu.org/quagga/docs/docs-info.html#Zebra-Protocol) they developed. 

I think that's a good model up to a point - I would like a watch dog process to serve a dedicated purpose, rather than rely on it to also serve such a central and critical role. But the idea of using basic RPC calls between processes would allow us to write very modular, specialized components, and would ensure that we can keep things simple from a programmable perspective. I'm thinking we still provide some level of programmatic access to each daemon for sure, but in a sunny day scenario, all programmatic interactions go through this abstraction layer - so you can retrieve all manner of configuration and state data from one place.

That said, each daemon will assuredly need to do certain things concurrently, so of course we will use goroutines, channels, etc. as normal within a running daemon.

**Comment by dave-tucker** [comment]
> I really believe the design laid out in Quagga is the right way to go - limiting potential failure domains by running important processes as truly separate processes. A failure in BGP wouldn't necessarily bring down the OSPF process, etc. Plus, we'd need a watchdog process anyways

Both approaches (process + threads, watchdog + individual processes) can work equally as well. What you've described above could easily happen in both approaches :)

> The way it's laid out in Quagga is that the watchdog process (zebra) is also the abstraction layer between the routing daemons and the kernel. So the daemons don't need to interact with the kernel directly, they make calls to zebra using a simple protocol they developed.

If each process needs to talk to the kernel we're doing it wrong. A protocol should be running a computation to calculate the routes it wants to offer for RIB. Our routing daemon has a rating process (think admin distance) to rate the routes before choosing what to install in the kernel (FIB)

> I would like a watch dog process to serve a dedicated purpose, rather than rely on it to also serve such a central and critical role. But the idea of using basic RPC calls between processes would allow us to write very modular, specialized components, and would ensure that we can keep things simple from a programmable perspective. I'm thinking we still provide some level of programmatic access to each daemon for sure, but in a sunny day scenario, all programmatic interactions go through this abstraction layer - so you can retrieve all manner of configuration and state data from one place.

Writing each protocol as a "library" and providing a well defined API so they can be consumed by the main process doesn't preclude the architecture that you speak of.

I'm conscious of creating protocols for protocols sake. The only reason I can think of to use RPC would be that would be planning on running the components on separate machines, or would want to provide an API such that you can write components in a different language (other than Go). We also have other fun problems to deal with: e.g what happens when an RPC doesn't get delivered (e.g process has hung or been stopped). Are we expected to replay messages? etc.. Next thing we know, we've started using ZeroMQ!

**Comment by bigmstone** [comment]
Only reason I would like to see ad-hoc processes is what if you want to consume only the BGP daemon. For instance if I wanted to create a overlay manager that distributed MAC stage via EVPN I care very little about OSPF and RIP.

However, if BGP is a library all its own it's easier for me to port into my application so that has more appeal.

Thoughts?

**Comment by dave-tucker** [comment]
So what if gopher-net has a TOML config file, where you can specify what protocols you want to run (and config details for each of those protocols)? Then your application can integrate over our RESTful API. Then if you wanted tighter integration, you can use the library directly to spawn your own daemon

**Comment by bigmstone** [comment]
I agree with that direction.

**Comment by Mierdin** [comment]
In that case, I would love to see some literature on process design in golang - what's conventional, some of the technical reasoning behind it. Anyone know where to point me? I think it will help me understand what the best practice is, instead of trying to apply generic principles I have been inheriting in the past 4 months :)

**Comment by nerdalert** [comment]
lmao at too much awesome. RE: config that was totally from OSRG peeps, (also noticed evpn recently). I thought it was super clever how the generate it. I need to poke around, I wonder if there is a library that will generate APIs from the config primitives. FSM is busted in a couple spots from channel n00b on my end. Generic MAC and/or IP notification on endpoints via EVPN with some data along with it would be useful to IaaS/container/NFV-(micro-services).

**Comment by Mierdin** [comment]
> Generic MAC and/or IP notification on endpoints via EVPN with some data along with it would be useful to IaaS/container/NFV-(micro-services).

I think we should be very particular around handling configuration vs state data, up to and including considering a separate mechanism for each. Existing network APIs do a fundamentally terrible job at this.

That said, state data should be a first-class citizen - obviously that's one of the most valuable things in all of this effort.

EDIT: On re-reading that I'd like to also mention this is a definite use case for separating out the two types of data. REST may be appropriate for config (or maybe not), but if a client wants to subscribe to the "state" API, it would be useful to have another mechanism. Think "OVSDB Monitor" but for MAC or IP data.

**Comment by bigmstone** [comment]
We should meaningfully abstract both sets of data, yes, and offer a trivial way of extracting the data in the method the consumer chooses. I would advocate for using something modeled in YANG since it fits well into the rest of the networking ecosystem. Think ODL. At least for a default and included way of extracting/pushing state. 

Perhaps channels could be used to stream both sets of data. Again. Thoughts?

**Comment by Mierdin** [comment]
> Think ODL.

No.

> Perhaps channels could be used to stream both sets of data. Again. Thoughts?

Yes, channels would likely need to be used to get data from a goroutine to whichever is handling the API layer, but I'm wondering how that data might be accessed from the outside. My first thought was some kind of message queue, but the senior developer on my team keeps popping in my head saying "DONT PUT EVERYTHING IN THE QUEUE". So I'm wondering what other "subscription" models exist for this kind of thing. Frankly despite the aforementioned voice in my head, the idea of sending state data into something that others can subscribe to is pretty sexy.

I'm going to show my ignorance here, but could we feed state data into something like etcd or consul?

**Comment by bigmstone** [comment]
I disagree with exporting it to a MQ ourselves. A MQ (or not) is the decision of the consumer and doesn't pass for an API. REST is an obvious choice, but again. It's a network stack. So something modeled in YANG will help it fit into existing tool chains. RESTCONF perhaps?

Again. If the user wants to attach the data to a MQ they can listen to the channel and push it to the queue themselves. That's not the job of gopher-net IMO.

**Comment by dave-tucker** [comment]
I guess it's best to understand the requirement  fully before making a technology choice re: data extraction. What are the use cases? 

IMO, REST is pretty poor for configuration. A a TOML-file would be much better and would be more UNIX'y (and most network folks I know have some degree of familiarity with this). E.g It's a config file, just like the router has a config file today. If it ain't broke... :)

For operational state, we need a better means of extraction, but this really varies on the type of state you are looking for. Do we want REST? a notifications API?, webhooks?, streaming data api? something else? I don't think the answer is that a single one will fit.. I think we'll end up with "best api for a specific type of data"

**Comment by Mierdin** [comment]
> IMO, REST is pretty poor for configuration. A a TOML-file would be much better and would be more UNIX'y (and most network folks I know have some degree of familiarity with this). E.g It's a config file, just like the router has a config file today. If it ain't broke... :)

Wholeheartedly agree wrt REST. The only caveat I would give is that we should definitely provide some kind of API-based mechanism for changing configuration, and treat it as a first-class citizen. The worst thing about Quagga, etc. is that it works on changing config files and restarting services. We should be able to very easily support the ability to make live changes via the API, and save those changes off to the config file. Basically supporting both methods.

> For operational state, we need a better means of extraction, but this really varies on the type of state you are looking for. Do we want REST? a notifications API?, webhooks?, streaming data api? something else? I don't think the answer is that a single one will fit.. I think we'll end up with "best api for a specific type of data"

Again, showing my ignorance here (since I don't really know the options) but something that provides push notifications would be ideal. I would love to have some kind of client that plugs into this thing, receives the initial state table(s) (with appropriate filters of course), and after that receives updates. Again, basically the OVSDB Monitor method but for MAC and IP info.

I'd love to hear some ideas for API designs that fit this model - I'm a blank canvas at this point.
